Hooking for MTLDevice basic new Library APIs

Zorro666 opened the pull request:
## Description

Hooking and serialization for MTLDevice basic new Library APIs

MTLDevice::newDefaultLibrary
MTLDevice::newLibraryWithSource

The hooked APIs return wrapped MTLLibrary objects to the application

Includes
* basic serialization support methods for MTLDevice
* basic MetalResourceManager implementation

Zorro666 committed:
Add support for NS::String* serialisation

Zorro666 committed:
Added ObjC::Get_defaultLibraryData

Zorro666 committed:
Add MetalResourceRecord record to WrappedMTLObject

Zorro666 committed:
Added ResourceId GetResID(WrappedMTLObject *obj)

Zorro666 committed:
Added basic MetalResourceManager implementation

Added MetalResourceManager to WrappedMTLDevice

Added helper macros:
METAL_NOT_IMPLEMENTED
METAL_NOT_IMPLEMENTED_ONCE

Added WrappedMTLDevice InitialState methods to metal_init_state.cpp
Added WrappedMTLDevice Serialiser helper methods to metal_core.cpp

Zorro666 committed:
Hooking for MTLDevice basic new Library APIs

Zorro666 commented on renderdoc/driver/metal/metal_device.cpp:
```diff
@@ -38,3 +51,121 @@ MTL::Device *WrappedMTLDevice::MTLCreateSystemDefaultDevice(MTL::Device *realMTL
+                                                             WrappedMTLLibrary *library,
+                                                             const void *pData, uint32_t bytesCount)
+{
+  SERIALISE_ELEMENT_LOCAL(Library, GetResID(library)).TypedAs("MTLLibrary"_lit);
```
Is this the correct approach or should I serialization support to WrappedMTLLibrary and its implementation would serialize its resource ID ?

baldurk replied:
In general case when referring to existing objects, yes this should be serialised via its ResourceId as a native serialisable type. However for creation functions there's a chicken-and-egg problem - on replay we haven't created the object yet because this is the function where it's created, so the object can't be serialised that way. Instead for create-time we do serialise the ResourceId in the way you have here so that's fine.

baldurk commented on renderdoc/driver/metal/metal_core.cpp:
```diff
@@ -0,0 +1,62 @@
+{
+  // check if a frame capture section version is supported
+  static const uint64_t CurrentVersion = 0x1;
+};
```
This struct should be in a header, as you'll need to access it from replay source files later.

Zorro666 replied:
moved to metal_core.h : 23d1d10f5bd213ce78d9c9ba85c6bce31927bcbb

Zorro666 replied:
89308d24b1b1cb2ef82dcfc3bf20077f5891f196

baldurk commented on renderdoc/driver/metal/metal_device.cpp:
```diff
@@ -23,12 +23,25 @@
+  GetResourceManager()->AddCurrentResource(objId, this);
+}
+
+void WrappedMTLDevice::Construct()
```
This is a strange pattern, why split out some of the constructor into this function? is there going to be another constructor in future, and if so why?

Zorro666 replied:
It is used in the future for the replay ie in

```
ReplayStatus Metal_CreateReplayDevice(RDCFile *rdc, const ReplayOptions &opts, IReplayDriver **driver)
{
...
  WrappedMTLDevice *wrappedMTLDevice = new WrappedMTLDevice();

  MetalReplay *replay = wrappedMTLDevice->GetReplay();
  replay->SetProxy(isProxy);
  ReplayStatus status = wrappedMTLDevice->Initialise(initParams, ver, opts);
...
}
```
and the `Initialise` method creates the `real` `MTLDevice`

That could be re-organized to do the `MTLDevice` creation in `Metal_CreateReplayDevice` and pass it to `WrappedMTLDevice::WrappedMTLDevice(MTL::Device *realMTLDevice, ResourceId objId)` and the remove the requirement for `WrappedMTLDevice::WrappedMTLDevice()` and `void WrappedMTLDevice::Construct()` ?

baldurk replied:
Yes the idea of the `InitParams` is that this struct can be used for any parameters 'outside' the normal flow of serialised API calls, those on the initial function or calls used to bootstrap the API in the first place. It is then serialised without any of the wrappers inside the `CreateReplayDevice` function, which uses the parameters to initialise the API and it would then pass the created metal device into the `WrappedMTLDevice` constructor.

You could organise it such that the `WrappedMTLDevice` creates itself in `Initialise()` which is what vulkan does, but in that case you should change it so the constructor never takes any parameters and the function call to create a device is a member function. It doesn't really fit with the more OO design of metal and how it organises its functions, so I'd suggest sticking closer to how D3D11/D3D12 organise as above.

In practice it doesn't look like metal has many parameters for device creation? I don't see any way of specifying a version etc so I assume it just always works at the newest version. It does look like you might want at minimum some kind of device identifier so you can pick the matching device in any system with multiple GPUs for `MTLCopyAllDevices` (though I don't know if in practice there exist any apple devices that have multiple GPUs so that might be an entirely theoretical concept?). This may become kind of a trivial distinction.

Zorro666 replied:
1f9a217577fe36c8bc218da0bbb2f3e721f22011

baldurk commented on renderdoc/driver/metal/metal_types.cpp:
```diff
@@ -0,0 +1,50 @@
+template <typename SerialiserType>
+void DoSerialise(SerialiserType &ser, NS::String *&el)
+{
+  const char *cStr = el ? el->utf8String() : NULL;
```
You should construct an `rdcstr` here. If you construct a C string the serialiser will have to cache it for its entire lifetime (since it doesn't know what the lifetime of the string is, it ensures it is valid indefinitely). If you serialise via an `rdcstr` it will just be a temporary for the sake of serialisation.

Zorro666 replied:
Thank you for a good spot. As the c-string is being copied into an NS::String could/should have been freeing the data anyway.

Something like this?
```
template <typename SerialiserType>
void DoSerialise(SerialiserType &ser, NS::String *&el)
{
  rdcstr rdcStr;
  if(el)
  {
    rdcStr = el->utf8String();
  }
  DoSerialise(ser, rdcStr);

  if(ser.IsReading())
  {
    el = NS::String::string(rdcStr.data(), NS::UTF8StringEncoding);
  }
}
```

After the change

<img width="473" alt="image" src="https://user-images.githubusercontent.com/39392/158048766-81f01931-121d-49bb-bda4-ee5040760df4.png">

baldurk replied:
Yep that's exactly what I had in mind.

Zorro666 replied:
d62879c53e9e464f4ba383b323c9783e2e056eec

Zorro666 replied:
9d0656f5d31a88dbff9bd5f3cd344699e45f8a56

baldurk commented on renderdoc/driver/metal/metal_types.cpp:
```diff
@@ -0,0 +1,50 @@
+void DoSerialise(SerialiserType &ser, NS::String *&el)
+{
+  const char *cStr = el ? el->utf8String() : NULL;
+  SERIALISE_ELEMENT(cStr);
```
Instead of using `SERIALISE_ELEMENT` which will treat NSString like a struct with a string member, you should forward the call on to `DoSerialise(ser, str)`. That will serialise "as-if" it were equivalent, so the NSString itself will be treated as just a string, not a struct.

See how interfaces are serialised 'as-if' they are a ResourceId.

Zorro666 replied:
Thanks yeah that looks a lot nicer resource inspector display

Zorro666 replied:
d62879c53e9e464f4ba383b323c9783e2e056eec

Zorro666 replied:
9d0656f5d31a88dbff9bd5f3cd344699e45f8a56

baldurk commented on renderdoc/driver/metal/metal_manager.h:
```diff
@@ -0,0 +1,168 @@
+  }
+  // ResourceManager interface
+
+  // template magic voodoo to unwrap types
```
You'll want these helpers at global scope so that Unwrap() and so on are available in any code, not just inside the resource manager.

Zorro666 replied:
89308d24b1b1cb2ef82dcfc3bf20077f5891f196

baldurk commented on renderdoc/driver/metal/metal_device.cpp:
```diff
@@ -38,3 +51,121 @@ MTL::Device *WrappedMTLDevice::MTLCreateSystemDefaultDevice(MTL::Device *realMTL
+{
+  MTL::Library *realMTLLibrary;
+
+  SERIALISE_TIME_CALL(realMTLLibrary = Unwrap<MTL::Device *>(this)->newDefaultLibrary());
```
You should improve the `Unwrap` template so that it type detects the correct unwrap, rather than needing an explicit template parameter like this.

baldurk commented on renderdoc/driver/metal/metal_helpers_bridge.mm:
```diff
@@ -0,0 +1,41 @@
+#import <AppKit/AppKit.h>
+#import <Foundation/NSStream.h>
+
+void ObjC::Get_defaultLibraryData(void *&pData, uint32_t &bytesCount)
```
I thought that metal-cpp meant there wasn't a need to jump into Obj-C like this? is there no metal-cpp equivalent of this call?

Zorro666 replied:
Unfortunately, this isn't covered by metal-cpp. 
This isn't Metal API, it is Foundation API which is Objective C only.
This method is similar to things like
`void getMetalLayerSize(void *layerHandle, int &width, int &height)` which is in `vk_apple_helpers.mm `or the helper methods in `apple_helpers.mm`

Could move this helper method (and there will be a few more in future PRs) into one of the existing objective C files ie. `apple_helpers.mm` ?

baldurk replied:
Oh I see. This is probably fine then - I'm presuming obj-c won't support the `bytebuf` type which would be preferred for use here instead of a raw `void*`/`uint32`? If not, then this is fine as-is.

Zorro666 replied:
I will investigate bytebuf : it should be possible to construct a bytebuf from the obj-c data in this method

JAKE :TODO:

Zorro666 replied:
That looks a lot cleaner

3a3c60864d1f66100080999ace37daf29d55459b

baldurk commented on renderdoc/driver/metal/metal_device.cpp:
```diff
@@ -38,3 +51,121 @@ MTL::Device *WrappedMTLDevice::MTLCreateSystemDefaultDevice(MTL::Device *realMTL
+  MTL::Library *realMTLLibrary;
+
+  SERIALISE_TIME_CALL(realMTLLibrary = Unwrap<MTL::Device *>(this)->newDefaultLibrary());
+  MetalResourceManager::UnwrapHelper<MTL::Library *>::Outer *wrappedMTLLibrary;
```
This kind of template instantiation makes it harder than necessary to see what the type is - it should instead just be `WrappedMTLLibrary`.

Zorro666 replied:
1f9a217577fe36c8bc218da0bbb2f3e721f22011

baldurk commented on renderdoc/driver/metal/metal_device.cpp:
```diff
@@ -38,3 +51,121 @@ MTL::Device *WrappedMTLDevice::MTLCreateSystemDefaultDevice(MTL::Device *realMTL
+}
+
+template <typename SerialiserType>
+bool WrappedMTLDevice::Serialise_mtlDevice_newDefaultLibrary(SerialiserType &ser,
```
By convention I always organise functions with the `Serialise_` version immediately before the real implementation. No particular reason, but it would be better to follow that convention.

Similarly, the `mtlDevice` is redundant in this function name because this is already implemented in `WrappedMTLDevice` so it should be `WrappedMTLDevice::Serialise_newDefaultLibrary`.

In general it's preferred to have serialise functions have the same signature as the function they are serialising, unless there's a strong reason to deviate. Cases like this where the return value needs to be serialised in some way would be one case, but in general where some parameters are implicit instead of explicit and need to be fetched like the pData here, it would be preferable to fetch them inside this function ahead of the serialisation in an `if(ser.IsWriting())` block, unless they are only fetchable in the outside real wrapper function.

I'm unclear why the `device` parameter is here. It seems like that is always going to be `this`? For the device where we only have one, it can be omitted entirely, for other functions where you need to serialise the current object you can serialise `this`.

Zorro666 replied:
Now the ObjC helper APIs are not needed and call directly to the real APIs from C++. This can be updated and follow existing patterns using helper macros like this

#define IMPLEMENT_FUNCTION_SERIALISED(ret, func, ...) \
  ret func(__VA_ARGS__);                              \
  template <typename SerialiserType>                  \
  bool CONCAT(Serialise_, func(SerialiserType &ser, __VA_ARGS__));

#define INSTANTIATE_FUNCTION_SERIALISED(CLASS, ret, func, ...)                                    \
  template bool CLASS::CONCAT(Serialise_, func(ReadSerialiser &ser, __VA_ARGS__)); \
  template bool CLASS::CONCAT(Serialise_, func(WriteSerialiser &ser, __VA_ARGS__));

The intent of the naming was to have the serialization method name match the chunk name and because the APIs are OO and not global wanted to include the name of the parent object in the chunk name just in case there is an API name that is the same in two different objects (say perhaps a common API name like "update"). Will keep the chunk name being unique and update the serialization macros and usage to match more existing patterns and remove unused parameters

baldurk replied:
Yes the chunk can have the object namespace e.g. `MetalChunk::Device_newDefaultLibrary` or similar, but it can be omitted from the function name itself as the namespacing already exists in the definition. Good to know this can be refactored then :+1:.

Zorro666 replied:
1f9a217577fe36c8bc218da0bbb2f3e721f22011

baldurk commented on renderdoc/driver/metal/metal_device.cpp:
```diff
@@ -38,3 +51,121 @@ MTL::Device *WrappedMTLDevice::MTLCreateSystemDefaultDevice(MTL::Device *realMTL
+  return true;
+}
+
+MTL::Library *WrappedMTLDevice::CreateMTLLibrary(const void *pData, uint32_t bytesCount)
```
Is this intended as a wrapper for `newLibrary`? It should share the same name then and (as closely as possible) the same function signature. This does seem like a bit of an interop challenge though.

Zorro666 replied:
It is a helper method for `Serialise_newDefaultLibrary` on replay and is not required for anything else.

It comes from before metal-cpp where this was implemented in the objective c helper file.

I will delete it from this PR as it is Replay orientated

Zorro666 replied:
1f9a217577fe36c8bc218da0bbb2f3e721f22011

baldurk reviewed (changes requested):
Some of the serialisation stuff doesn't follow the normal conventions which I've made more specific comments on. This might come out in the wash in future revisions it's unclear yet to me from this how much is necessary to workaround Obj-C and how much can be improved.

Zorro666 committed:
Change ObjCWrappedMTL* to ObjCBridgeMTL*

Change:
AllocateObjCWrapper -> AllocateObjCBridge
IsObjCWrapped -> IsObjCBridge
GetObjC -> GetObjCBridge
WrappedMTLObject::objc -> WrappedMTLObject::objcBridge
WrappedMTLObject::GetObjCWrappedMTLDevice -> WrappedMTLObject::GetObjCBridgeMTLDevice

Added GetReal() to C++ Wrapped objects

Zorro666 committed:
Use bytebuf to serialise default library data

Zorro666 committed:
Remove extra #include "serialise/serialiser.h"

Zorro666 committed:
Add declarations for ObjC helper methods

baldurk commented on renderdoc/driver/metal/metal_types.h:
```diff
@@ -24,16 +24,30 @@
+  extern WrappedMTL##CPPTYPE *GetWrapped(MTL::CPPTYPE *objCWrapped); \
+  extern MTL::CPPTYPE *GetReal(MTL::CPPTYPE *objCWrapped);           \
+  extern bool IsObjCWrapped(MTL::CPPTYPE *objCWrapped);              \
+  extern ResourceId GetId(MTL::CPPTYPE *objCWrapped);                \
```
Minor nitpick, this free function should be `GetResID` to match the other backends and be consistent. There's already a `GetResID` for the generic `WrappedMTLObject`, this would just be an overload which works on the unwrapped type to look up via a cast/GetWrapped call.

baldurk reviewed (changes requested):
One more minor naming issue with an added helper, but otherwise this looks good to me now.

Zorro666 committed:
Rename ObjC helper GetId -> GetResId

Zorro666 committed:
Rename ObjC helper GetResId -> GetResID
